pkg/policy/mapstate.go

```diff
@@ -498,24 +629,198 @@ func (keys MapState) denyPreferredInsertWithChanges(newKey Key, newEntry MapStat
+				if k.DestPort == 0 && other.DestPort != 0 {
+					return 2 // */proto/* is a superset of */proto/port
+				} // else more specific or different ports
+			} // else more specific or different protocol
+		} else {
+			// Wildcard L3 is a superset of a specific L3 only if wildcard L3 is also wildcard L4, or the L4's match between the keys
+			if k.Nexthdr == 0 { // k.DestPort == 0 is implied
+				return 1 // */*/* is a superset of ID/x/x
+			} else if k.Nexthdr == other.Nexthdr {
+				if k.DestPort == 0 {
+					return 2 // */proto/* is a superset of ID/proto/x
+				} else if k.DestPort == other.DestPort {
+					return 3 // */proto/port is a superset of ID/proto/port
+				} // else more specific or different ports
+			} // else more specific or different protocol
+		}
+	} else if k.Identity == other.Identity {
+		if k.Nexthdr == 0 {
+			if other.Nexthdr != 0 {
+				return 4 // ID/*/* is a superset of ID/proto/x
+			} // else both are ID/*/*
+		} else if k.Nexthdr == other.Nexthdr {
+			if k.DestPort == 0 && other.DestPort != 0 {
+				return 5 // ID/proto/* is a superset of ID/proto/port
+			} // else more specific or different ports
+		} // else more specific or different protocol
+	} // else more specific or different identity
+	return 0
+}
+
+// authPreferredInsert applies AuthType of a more generic entry to more specific entries, if not
+// explicitly specified.
+func (keys MapState) authPreferredInsert(newKey Key, newEntry MapStateEntry, features policyFeatures, changes ChangeState) {
+	if features.contains(authRules) {
+		if newEntry.hasAuthType == DefaultAuthType {
+			// New entry has a default auth type.
+			// Fill in the AuthType from more generic entries with an explicit auth type
+			max_specificity := 0
+			l3l4keys := make(MapState)
+			for k, v := range keys {
+				// Only consider the same Traffic direction
+				if newKey.TrafficDirection != k.TrafficDirection {
+					continue
+				}
+				// Skip deny entries, they have no effect on authentication requirements,
+				// and we should not create any new (non-deny) entries based on a combination
+				// of an existing deny and the new non-deny entry.
+				if v.IsDeny {
+					continue
+				}
+
+				// Nothing to be done if entry has default AuthType
+				if v.hasAuthType == DefaultAuthType {
+					continue
+				}
+
+				// Find out if 'k' is a more specific superset of 'newKey' than seen so far
+				if specificity := k.IsSuperSetOf(newKey); specificity > 0 {
+					if specificity > max_specificity {
+						newEntry.AuthType = v.AuthType
+						max_specificity = specificity
+					}
+				} else {
+					// Check if a new L3L4 entry must be created due to L3-only
+					// 'k' specifying an explicit AuthType and an L4-only 'newKey' not
+					// having an explicit AuthType. In this case AuthType should
+					// only override the AuthType for the L3 & L4 combination,
+					// not L4 in general.
+					//
+					// These need to be collected and added after it is known if
```

Comment:
nit, I think this might improve clarity? Feel free to take or skip:

```suggestion
					// These need to be collected and only added if
```